### PR TITLE
Rewrite in TypeScript

### PR DESCRIPTION
**Pull request by JunichiSugiura**
#18 

@jescalan It seems like TypeScript doesn't currently support [subpaths exports](https://nodejs.org/docs/latest-v14.x/api/esm.html#esm_subpath_exports) as described [here](https://github.com/Microsoft/TypeScript/issues/8305) and [here](https://github.com/microsoft/TypeScript/issues/33079).

One solution is to export everything from `index.ts`. However it'll introduce breaking changes since every consumer has to change the way to import.

```ts
// currently
import hydrate from 'next-mdx-remote/hydrate'
import renderToString from 'next-mdx-remote/render-to-string'

// if we export all from `index.ts`
import {  hydrate, renderToString } from 'next-mdx-remote'
```

This is not a big library so the change above is acceptable I think but it's up to how you feel about the api change.
Let me know what you think.

### CODE REVIEW

**Comment by JunichiSugiura** [comment]
Sounds great! Let me find a time to work on it tomorrow.

**Comment by jescalan** [comment]
Ok I have received another update that it's "not recommended", and should be tested to make sure it works haha

**Comment by JunichiSugiura** [comment]
Sorry, it's been a crazy week for me. I'll reschedule it on Saturday.

**Comment by JunichiSugiura** [comment]
Hmm, I've tested with re-exporting both `renderToString` and `hydrate` from `index.tsx` but it throw "fs" module not found error on the next app side. Which indicates tree shaking doesn't work and `renderToString` is still included inside client side code. I've also tested with next@9.5 (webpack5 beta enabled) but no luck.

**Comment by JunichiSugiura** [comment]
I found out if I respect current file structure (i.e. `hydrate.ts` instead of `src/hydrate.ts`), TS resolves submodules on the app site. I'll remove `src/` and `dist/`.

**Comment by JunichiSugiura** [comment]
@jescalan It should do the trick now.

File: __tests__/index.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ const puppeteer = require('puppeteer')
 const handler = require('serve-handler')
 const http = require('http')
 const rmfr = require('rmfr')
-const renderToString = require('../render-to-string')
+const renderToString = require('../render-to-string').default
```
Comment (jescalan) [inline __tests__/index.js:8]: why is this necessary now?

File: babel-plugin-mdx-browser.js [hunk 2]
```diff
@@ -2,17 +2,16 @@ module.exports = function BabelPluginMdxBrowser() {
...
     visitor: {
       // remove all imports, we will add these to scope manually
-      ImportDeclaration(path) {
+      ImportDeclaration: function (path) {
         path.remove()
       },
       // the `makeShortcode` template is nice for error handling but we
       // don't need it here as we are manually injecting dependencies
-      VariableDeclaration(path) {
+      VariableDeclaration: function (path) {
```
Comment (jescalan) [inline babel-plugin-mdx-browser.js:10]: any reason for these additions?

File: package.json [hunk 3]
```diff
@@ -9,13 +9,17 @@
   "dependencies": {
     "@babel/core": "^7.10.5",
     "@babel/preset-env": "^7.10.4",
+    "@babel/preset-react": "^7.10.4",
```
Comment (jescalan) [inline package.json:12]: 🤦🏼‍♂️ good catch here

File: render-to-string.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import mdx, { Options } from '@mdx-js/mdx'
+import { MDXProvider, mdx as mdxReact } from '@mdx-js/react'
+import { transformAsync } from '@babel/core'
+import presetEnv from '@babel/preset-env'
+import presetReact from '@babel/preset-react'
+import { renderToString as reactRenderToString } from 'react-dom/server'
+import React from 'react'
+// @ts-ignore
```
Comment (jescalan) [inline render-to-string.ts:8]: Just curious, why ignore this?

File: render-to-string.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    transformAsync(code, {
+      presets: [presetReact, presetEnv],
+      plugins: [pluginBrowser],
+      configFile: false,
+    }),
+  ])
+
+  if (!now || !later || !later.code) {
+    throw new Error('Failed to transform mdx source code')
+  }
```
Comment (jescalan) [inline render-to-string.ts:36]: This doesn't seem like a super useful error message - do you have any more context on when you encountered this and what caused it? Maybe we can add more detail to the error?

File: types.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import { createElement } from 'react'
+
+export interface Scope {}
+
+export type Components = Parameters<typeof createElement>[1]
```
Comment (jescalan) [inline types.ts:5]: This is some crazy wizardry here ✨ - could you help me to understand what's going on? afaik, the `Parameters` utility creates a type out of the arguments to a function, but you're passing in the results of a `typeof` operation which I'd think would return `"function"`?

**Comment by jescalan** [comment]
Just pulled this down to run tests and see if we could get it live - both mdx and babel-preset-react do not have typescript types, and types are not present in the library or definitely-typed for either one. `tsc` errors on `npm test` due to `implicitAny` being set to false.

I tinkered for a bit with trying to get types in for these libraries but ultimately moved on, its not worth the effort for me right now. Let me know if you saw this during your own local dev and how you might want to handle it!

**Comment by brianespinosa** [comment]
Well... @JunichiSugiura you went above and beyond on what I did in my PR. I guess I should have looked to see if any PRs were open doing something similar. 👍🏽  This PR would resolve #24 and then some.

@jescalan let me know if I should decline #26 and instead submit a separate PR for #25

File: types.ts (hunk 6 above)
Comment (zefexdeveloper) [reply types.ts:5]: @jescalan No, the `typeof` here means the types of the parameters.

For example:

```ts
declare function f1(arg: { a: number; b: string }): void;

type T3 = Parameters<typeof f1>;
//    ^ = type T3 = [arg: {
    a: number;
    b: string;
}]
```

Comment (zefexdeveloper) [reply types.ts:5]: @jescalan No, the `typeof` here means the types of the parameters.

For example:

```ts
declare function f1(arg: { a: number; b: string }): void;

type T3 = Parameters<typeof f1>;
//    ^ = type T3 = [arg: {
    a: number;
    b: string;
}]
```

File: babel-plugin-mdx-browser.js (hunk 2 above)
Comment (zefexdeveloper) [reply babel-plugin-mdx-browser.js:10]: > any reason for these additions?

I don't see a reason for these changes.

File: __tests__/index.js (hunk 1 above)
Comment (zefexdeveloper) [reply __tests__/index.js:8]: > why is this necessary now?

CommonJS (Node's require() and exports) is not compatible with ES6 default exports. But in this case as we are talking about puppeteer tests I don't see a reason. We are not using TS here.

**Comment by zefexdeveloper** [comment]
@jescalan I answered a few questions you had on his code.

**Comment by JunichiSugiura** [comment]
I'm sorry it's been a crazy few weeks for me. Let me get back to you tomorrow.

**Comment by miku86** [comment]
> I'm sorry it's been a crazy few weeks for me. Let me get back to you tomorrow.

Hello Junichi,

thanks for your work so far.

Any help required to finish this?

File: __tests__/index.js (hunk 1 above)
Comment (JunichiSugiura) [reply __tests__/index.js:8]: > CommonJS (Node's require() and exports) is not compatible with ES6 default exports

Yes, that's the reason. Without it'll throw `TypeError: renderToString is not a function`

Comment (JunichiSugiura) [reply __tests__/index.js:8]: We can also convert all the test files to TS but I don't have enough time for it.

File: render-to-string.ts (hunk 4 above)
Comment (JunichiSugiura) [reply render-to-string.ts:8]: Because the babel plugin module is written in JS and we don't have type definition for it, tsc complains.
I'm not sure where to find type definitions for babel plugins. Maybe someone can help.

File: render-to-string.ts (hunk 5 above)
Comment (JunichiSugiura) [reply render-to-string.ts:36]: The return type of `transformAsync` is `Promise<BabelFileResult | null>` and we need to handle the null case. Which is unlikely to happen since it only depends on babel plugins and `code` const which is returned by `mdx(...)`.

File: render-to-string.ts (hunk 4 above)
Comment (JunichiSugiura) [reply render-to-string.ts]: I provided `babel-plugin-mdx-browser.d.ts`: https://github.com/hashicorp/next-mdx-remote/pull/20/commits/1341796813f82d61216c422cabf6de804f74780f

**Comment by JunichiSugiura** [comment]
> tsc errors on npm test due to implicitAny being set to false

It's due to `@types/` which I *had* locally wasn't committed. I stupidly run `git clean -xdf` to reproduce your error and now all the definition files inside the dir are gone🤦‍♂️

I'll recreate these files.

**Comment by JunichiSugiura** [comment]
Sorry for taking it so long. I finally managed to find some time to
- reflect some reviews
- fix third-party related type errors
- merge master

Hope it works this time🤞

**Review by jescalan — COMMENTED** [review]
Thank you again for the great work here @JunichiSugiura! We're getting super close I think

File: babel-plugin-mdx-browser.ts [hunk 7]
```diff
@@ -1,18 +1,17 @@
-module.exports = function BabelPluginMdxBrowser() {
+export default function BabelPluginMdxBrowser() {
   return {
     visitor: {
       // remove all imports, we will add these to scope manually
-      ImportDeclaration(path) {
+      ImportDeclaration(path: any) {
```
[the remark this answers is not in the document]
Comment (06kellyjac) [reply babel-plugin-mdx-browser.ts:5]: These are the types I got working (explanation in the commit):
https://github.com/06kellyjac/next-mdx-remote/commit/d5910b083be2a48790e5a7f3293d893bb1abc29b

I couldn't really find any docs for writing babel plugins in typescript :shrug: so I had to dive through the definitions.
Also it's probably worth double checking the [Optional Chaining](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining) logic matches before. + that the code removed was actually incorrect.

The tests ran successfully.

**Comment by jescalan** [comment]
I'm going to close this one due to inactivity, and because we have typings shipping out shortly
